### DISCUSSION TITLE
how to get request.files attribute after flask send_from_directory?

### DISCUSSION THREAD

**Discussion by dayyass** [pallets/flask#4707] [Q&A]
I have a master/slave architecture with master and slave flask restful apps and I want to send file from slave node to master node to send it further to user (different node).

I used send_from_directory function to do it, but after applying that function on slave node, the request on master node doesn’t have files attribute, so can neither push it forward to user node nor save it in tmp folder and send it to user from file system with send_from_directory.

Could anyone help me with my problem?
I’ve spent several days for that without any success.

**Comment by ThiefMaster** [comment]
The description of your architecture is not very clear....

**Reply by dayyass** [reply]
Not diving into business logic, there is a centralized distributed system with **one master node**, **several slave nodes** and **user nodes**. There is a user request to do something and GET requests goes to master node, then it searches for slave node and master node requests slave note to return file to push it back to user.

**USER_NODE** -> GET Request -> **MASTER_NODE** -> GET Request -> **SLAVE_NODE**
**USER_NODE** <- file <- **MASTER_NODE** <- file <- **SLAVE_NODE**

**Reply by ThiefMaster** [reply]
So if you create files on the slave nodes, then you need to somehow make them accessible to your master node. Either by using distributed storage (S3 or similar are a great option for this) or by sending them back / exposing them via HTTP

**Reply by dayyass** [reply]
I want to send it from slave to master via HTTP, and I used **send_from_directory** function on slave, but I couldn't get file on master from request

**Reply by epivoca** [reply]
Actually, I've faced the same problem and have been trying to solve it for a while, but still without any results. Would be really glad to find some solutions here!

**Reply by dayyass** [reply]
Is it possible to create your own Response() object and put opened files or werkzeug.FileStorage in its .files attribute?

**Comment by tachyondecay** [comment]
You seem to be confused about what `request.files` is for. This is for accessing files uploaded by the client to your app via a POST request.

You cannot "put" a file into `request.files`.

Based on your workflow description, your master node needs to take the response from the slave node as a `BinaryIO` stream and use `send_file()` to pass it on to the user.

This is a very inefficient way to deliver a file to a user. Storing your files in a common distributed solution, as ThiefMaster noted, will be more efficient and maintainable.

**Reply by epivoca** [reply]
Thank you for pointing out some mistakes.

 You have described how to take a response from the slave node, but our problem is on the stage of sending a file from slave node to master node, whereas we are using `send_file()`.

According to your vision of this process, can you please share with us an abstract code-example of sending file?

**Reply by davidism** [reply]
What you're describing has nothing to do with Flask at this point. If you need to move data from one server to another, you need to design that system. It's not related to HTTP requests and responses, which is what the Flask framework handles.
